global-script.lua

```diff
@@ -3245,4 +3410,282 @@ function tCompare(t1,t2)
+        return
+    end
+
+    readyTokens[fromColor], readyTokens[toColor] = readyTokens[toColor], readyTokens[fromColor]
+    selectedColors[fromColor], selectedColors[toColor] = selectedColors[toColor], selectedColors[fromColor]
+
+    -- Remove any items still in the bags
+    -- NOTE: TTS's documentation suggests we may need to wait a physics frame after doing this,
+    -- but this seems to work fine without doing that.  If something goes awry with this code in the future,
+    -- you might try re-adding the delay.
+    for color,data in pairs(colors) do
+        for i = 1,data.qty do 
+            local obj = data.bag.takeObject({
+                sound=false,
+                position={x=data.ix*2, z=200, y=i*2}    -- Chosen to be out-of-the-way and to prevent items from stacking.
+            })
+            table.insert(data.bagContents, obj)
+        end
+    end
+
+    -- Pass 1: Iterate over all objects looking for "<color>'s X".
+    -- Make a note of what we find and what tint it is.
+    local match = string.match  -- Performance
+    local name, suffix
+    for _,obj in pairs(getAllObjects()) do
+        name = obj.getName()
+        if name then
+            for _,data in pairs(colors) do 
+                suffix = match(name, data.pattern)
+                if suffix then
+                    data.tints[suffix] = obj.getColorTint()
+                    if not data.objects[suffix] then
+                        data.objects[suffix] = {obj}
+                    else
+                        table.insert(data.objects[suffix], obj)
+                    end
+                end
+            end
+        end
+    end
+
+    -- Pass 2: Iterate over found objects and swap color tints and object names.
+    -- After we're done, put objects in their new presence bag, if applicable.
+    for _,ab in pairs({{colors.from, colors.to}, {colors.to, colors.from}}) do
+        local a, b = unpack(ab)
+        for suffix, tint in pairs(a.tints) do
+            local newname = a.color .. "'s " .. suffix
+            for _, obj in ipairs(b.objects[suffix]) do
+                obj.setColorTint(tint)
+                obj.setName(newname)
+            end
+        end
+        for i = #b.bagContents,1,-1 do  -- Iterate in reverse order.
+            a.bag.putObject(b.bagContents[i])
+        end
+    end
+end
+
+
+function swapPlayerColors(a, b)
+    if a == b then return end
+    local pa, pb = Player[a], Player[b]
+
+    if not playerBlocks[a] then
+        -- This should only trigger if the player clicking is a non-standard color.
+        if pb.seated then
+            broadcastToColor("Color " .. b .. " is already claimed.  Try another color.", a, Color.Red)
+            return false
+        end
+    end
+
+    if pa.seated then
+        if pb.seated then
+            -- Both players seated.  Need an intermediary.
+            -- Hopefully nobody is... Pink?
+            pb.changeColor("Pink")
```

Comment:
`changeColor` returns a bool if the operation succeeded so this process should probably be aborted if that happens

Reply:
Latest fork now tries several player colors in order until one works, and complains/aborts if none do.  There's also a warning if swapping with another player in hotseat since it can break in interesting ways.